### PR TITLE
[Flax] Add scan_with_axes

### LINKED ISSUE #17399: RFC
### Feature request

We should add scan and remat (gradient checkpointing) to the most important Flax/JAX models (BERT, GPT2, OPT, T5, BART, Wav2Vec2).

### Motivation

Scan allows for much faster compilation and memory savings and `remat` is the equivalent of `gradient_checkpointing` in PyTorch.

@sanchit-gandhi already uses both features in the Flax Seq2Seq Speech project - see: https://github.com/sanchit-gandhi/seq2seq-speech so it'd be quite trivial to get them working.

**Implementation details:**

Given that both `scan` and `remat` are not related to the model architecture, they should IMO **not** be in the model's config (We've done this mistake in PyTorch and don't want to repeat it here).

I would advocate for the following API:

```python
model = FlaxBertForMaskedLM.from_pretrained("bert-base-cased")
model.scan()  # or model.scan_enable()
model.unscan()  # or model.scan_disable()
```

and 

```python
model = FlaxBertForMaskedLM.from_pretrained("bert-base-cased")
model.gradient_checkpoint_enable()
model.gradient_checkpoint_disable()
```

As can be seen here: https://github.com/sanchit-gandhi/seq2seq-speech/blob/b28d0c25c8fad0f9ffa6707f91f7aba320d44a4b/models/modeling_flax_wav2vec2.py#L504

We'll need to re-initialize the `flax.linen.module` inside the model. However this should be fine since it just means that we do

```
self.module = self.module_class(config=config, dtype=dtype, use_scan=True, **kwargs)
 self. _is_scan_enabled = True
```

similar to this line: https://github.com/huggingface/transformers/blob/71e602725b90f63f404109bae9f72cbdf755477b/src/transformers/models/wav2vec2/modeling_flax_wav2vec2.py#L868

We can see along the PR how much logic can reside in `modeling_flax_utils.py` and how much would go into the specific models, *e.g.* `modeling_flax_wav2vec2.py`.

The same API / logic could be used for the `gradient_checkpointing`.

### Your contribution

Happy to give this implementation a shot with @sanchit-gandhi  and @patil-suraj .

Also would love to hear feedback from @borisdayma @marcvanzee about the API

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Adds `scan_with_axes` to Flax Bert and its derived models.

TODO:
- [ ] Fix cookie cutter template
- [ ] Run `make fix-copies` (after review)

Fixes [#17399](https://github.com/huggingface/transformers/issues/17399)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sanchit-gandhi** [comment]
Current API:

1. With automatic init:
```python
model.scan_enable()  # to enable scan
model.scan_disable()  # to disable scan
```

2. Without automatic init:
```python
model.scan_enable()  # to enable scan in the nn.Module
params = model.convert_unroll_to_scan(params) # to convert the unrolled params to scan

model.scan_disable()  # to disable scan in the nn.Module (i.e. unrolled)
params = model.convert_scan_to_unroll(params) # to convert the scan params to unrolled
```

With automatic init, the params are converted from unrolled to scan under the hood ([`.convert_unroll_to_scan()`](https://github.com/huggingface/transformers/blob/98c42b01c2f1128b22bbda2cdc3d25e03b94af53/src/transformers/modeling_flax_utils.py#L435)). 

The params are converted to/from scan on the fly, with no memory overhead for conversion (_c.f._ [`.convert_unroll_to_scan()`](https://github.com/huggingface/transformers/blob/98c42b01c2f1128b22bbda2cdc3d25e03b94af53/src/transformers/modeling_flax_utils.py#L435))
```python
for i in range(self.config.num_hidden_layers):
    # Stack the params for the N layers into one super block
    # and remove the unrolled layer params on the fly
    # -> no memory overhead for conversion!
    unrolled_layer = params.pop(key.replace("0", str(i)))
    stacked_params.append(unrolled_layer)
```

**Comment by qgallouedec** [comment]
Closing, see https://github.com/huggingface/transformers/issues/17399#issuecomment-5016868117
